File: android/app/src/main/java/com/noop/analytics/PhysiologicalSteps.kt
package com.noop.analytics

import com.noop.data.StepSample

/**
 * WHOOP-style step-day boundaries. A physiological step day starts at the onset of the main sleep
 * and ends at the next main-sleep onset; the delta is still attributed to the later counter sample.
 * We deliberately do not mask the sleep span: a real walk during a long night-time wake remains a
 * walk, while the shared activity-class and tick-rate gates continue to reject non-locomotion.
 */
object PhysiologicalSteps {

    enum class SleepKind { MAIN_SLEEP, NAP, UNCLASSIFIED }

    data class SleepBlock(
        val onset: Long,
        val end: Long,
        /** Stable identity lets a user-edited onset move an existing boundary instead of adding one. */
        val id: String = onset.toString(),
        val editedOnset: Long? = null,
        val kind: SleepKind = SleepKind.UNCLASSIFIED,
    ) {
        val effectiveOnset: Long get() = editedOnset ?: onset
    }

    /** The one open physiological cycle. Absence of a newly detected night never creates a calendar cycle. */
    data class CycleBoundary(val sleepId: String, val onset: Long)

    data class CycleWindow(val sleepId: String, val onset: Long, val endExclusive: Long)

    data class OwnerSegment(val owner: String, val onset: Long, val endExclusive: Long)
    data class OwnerCoverage(
        val owner: String,
        val onset: Long,
        val endExclusive: Long,
        val priority: Int,
    )

    fun shouldReadCounterPredecessor(segmentIndex: Int): Boolean = segmentIndex == 0

    private const val MIN_MAIN_SLEEP_SECONDS = 3 * 3_600L

    /**
     * State carried between ascending database pages. [previous] is deliberately retained even when it
     * lies before the cycle so the first in-cycle counter delta has the correct predecessor.
     */
    data class CycleCountState internal constructor(
        val hasActivityClasses: Boolean,
        val previous: StepSample?,
        val total: Long,
        val evaluatedDelta: Boolean,
    )

    private fun isPlausibleMainSleep(
        block: SleepBlock,
        tzOffsetSeconds: Long,
        habitualMidsleepSec: Long?,
    ): Boolean {
        if (block.kind == SleepKind.NAP) return false
        val onset = block.effectiveOnset
        if (block.end <= onset) return false
        if (block.kind == SleepKind.MAIN_SLEEP) return true
        val midpoint = onset + (block.end - onset) / 2
        val target = SleepStageTotals.targetMidsleepSec(habitualMidsleepSec)
        return SleepStageTotals.circularDistanceSec(
            SleepStageTotals.localSecOfDay(midpoint, tzOffsetSeconds),
            target,
        ) < SleepStageTotals.ALIGNMENT_ZERO_SEC
    }

    /**
     * Apply the same user-visible main-vs-nap shape used by the sleep surfaces: only the canonical overnight
     * group can be MAIN, it must total at least three hours, and every other/twinless-explicit block is NAP.
     */
    fun classifyForCycle(
        blocks: List<SleepBlock>,
        tzOffsetSeconds: Long,
        habitualMidsleepSec: Long?,
    ): List<SleepBlock> {
        if (blocks.isEmpty()) return emptyList()
        val explicitMain = blocks.indices.filter { blocks[it].kind == SleepKind.MAIN_SLEEP }
        val selectable = blocks.indices.filter { blocks[it].kind != SleepKind.NAP }
        val selectedOriginalIndices = if (explicitMain.isNotEmpty()) {
            explicitMain
        } else {
            val selectableNightBlocks = selectable.map {
                val b = blocks[it]
                SleepStageTotals.NightBlock(b.effectiveOnset, b.end)
            }
            // Eliminate nap-shaped GROUPS before choosing a winner. Otherwise a six-hour afternoon nap can
            // win the generic duration scorer, fail the daytime guard, and hide a valid shorter night.
            val eligible = SleepStageTotals.bridgedNightGroups(selectableNightBlocks, tzOffsetSeconds)
                .filter { group ->
                    val total = group.indices.sumOf { i -> selectableNightBlocks[i].durationS.coerceAtLeast(0L) }
                    val onset = group.indices.minOfOrNull { selectableNightBlocks[it].start }
                    total >= MIN_MAIN_SLEEP_SECONDS && onset != null &&
                        SleepStageTotals.isOvernightOnset(onset, tzOffsetSeconds)
                }
                .flatMap { it.indices }
                .distinct()
            SleepStageTotals.mainNightGroupIndices(
                eligible.map { selectableIndex ->
                    val original = selectable[selectableIndex]
                    val b = blocks[original]
                    SleepStageTotals.NightBlock(b.effectiveOnset, b.end)
                },
                tzOffsetSeconds,
                habitualMidsleepSec,
            ).orEmpty().map { selectedIndex ->
                selectable[eligible[selectedIndex]]
            }
        }.toHashSet()
        return blocks.mapIndexed { index, block ->
            block.copy(kind = if (index in selectedOriginalIndices) SleepKind.MAIN_SLEEP else SleepKind.NAP)
        }
    }

    /** Use the exact same main-night/group selector as the Sleep screen and daily sleep aggregate. */
    fun mainSleepOnset(
        blocks: List<SleepBlock>,
        tzOffsetSeconds: Long,
        habitualMidsleepSec: Long?,
    ): Long? {
        val eligible = classifyForCycle(blocks, tzOffsetSeconds, habitualMidsleepSec)
            .filter { it.kind == SleepKind.MAIN_SLEEP }
        val indices = SleepStageTotals.mainNightGroupIndices(
            eligible.map { SleepStageTotals.NightBlock(it.effectiveOnset, it.end) },
            tzOffsetSeconds,
            habitualMidsleepSec,
        ) ?: return null
        return indices.minOfOrNull { eligible[it].effectiveOnset }
    }

    /**
     * Advance the single open cycle from newly observed sleep rows. Missing sleep and nap-only input keep
     * [previous] unchanged. A late-arriving main sleep advances it; a stable [SleepBlock.id] applies a
     * user-edited onset to the same boundary even when the edit moves backwards.
     */
    fun advanceBoundary(
        previous: CycleBoundary?,
        observedBlocks: List<SleepBlock>,
        now: Long,
        tzOffsetSeconds: Long = 0,
        habitualMidsleepSec: Long? = null,
    ): CycleBoundary? {
        val eligible = observedBlocks.filter {
            it.effectiveOnset <= now && isPlausibleMainSleep(it, tzOffsetSeconds, habitualMidsleepSec)
        }
        previous?.let { old ->
            eligible.firstOrNull { it.id == old.sleepId }?.let {
                val edited = CycleBoundary(it.id, it.effectiveOnset)
                val newer = eligible.filter { candidate -> candidate.id != old.sleepId }
                    .maxByOrNull { candidate -> candidate.effectiveOnset }
                return if (newer != null && newer.effectiveOnset > edited.onset) {
                    CycleBoundary(newer.id, newer.effectiveOnset)
                } else {
                    edited
                }
            }
        }
        val newest = eligible.maxByOrNull { it.effectiveOnset } ?: return previous
        return if (previous == null || newest.effectiveOnset > previous.onset) {
            CycleBoundary(newest.id, newest.effectiveOnset)
        } else {
            previous
        }
    }

    /**
     * Build historical sleep-to-sleep windows without inventing rows at midnight. The final observed
     * boundary alone owns the open tail to [now], including days on which no new sleep was detected.
     */
    fun cycleWindows(boundaries: List<CycleBoundary>, now: Long): List<CycleWindow> {
        val ordered = boundaries.asSequence()
            .filter { it.onset <= now }
            .distinctBy { it.sleepId }
            .sortedBy { it.onset }
            .toList()
        return ordered.mapIndexedNotNull { index, boundary ->
            val end = ordered.getOrNull(index + 1)?.onset ?: now
            if (end <= boundary.onset) null else CycleWindow(boundary.sleepId, boundary.onset, end)
        }
    }

    /** Split at local midnights and merge adjacent same-owner slices; counters never cross device seams. */
    fun ownerSegments(
        window: CycleWindow,
        ownerByDay: Map<String, String>,
        tzOffsetSeconds: Long,
        fallbackOwner: String,
    ): List<OwnerSegment> {
        if (window.endExclusive <= window.onset) return emptyList()
        val out = ArrayList<OwnerSegment>()
        var cursor = window.onset
        var lastOwner = fallbackOwner
        while (cursor < window.endExclusive) {
            val day = AnalyticsEngine.dayString(cursor, tzOffsetSeconds)
            val local = cursor + tzOffsetSeconds
            val dayNumber = Math.floorDiv(local, SleepStageTotals.SECONDS_PER_DAY)
            val dayStart = dayNumber * SleepStageTotals.SECONDS_PER_DAY - tzOffsetSeconds
            // The sleep-owner owns the partial onset day. Calendar owner changes take effect only at a
            // midnight seam, never retroactively before the boundary sleep began.
            val owner = if (cursor == window.onset && cursor > dayStart) {
                fallbackOwner
            } else {
                ownerByDay[day] ?: lastOwner
            }
            val nextMidnight = (dayNumber + 1) * SleepStageTotals.SECONDS_PER_DAY - tzOffsetSeconds
            val end = minOf(window.endExclusive, nextMidnight)
            val previous = out.lastOrNull()
            if (previous != null && previous.owner == owner && previous.endExclusive == cursor) {
                out[out.lastIndex] = previous.copy(endExclusive = end)
            } else {
                out += OwnerSegment(owner, cursor, end)
            }
            lastOwner = owner
            cursor = end
        }
        return out
    }

    /**
     * Timestamp coverage wins at every seam; priority resolves overlap, so two worn straps never add.
     * Each coverage is a continuous first..last ownership span by design: an internal sample/radio gap
     * creates no seam and therefore cannot hand the interval to a lower-priority concurrently worn strap.
     */
    fun ownerSegmentsFromCoverage(
        window: CycleWindow,
        coverage: List<OwnerCoverage>,
        fallbackOwner: String,
    ): List<OwnerSegment> {
        if (window.endExclusive <= window.onset) return emptyList()
        val clipped = coverage.mapNotNull {
            val start = maxOf(window.onset, it.onset)
            val end = minOf(window.endExclusive, it.endExclusive)
            if (end <= start) null else it.copy(onset = start, endExclusive = end)
        }
        val seams = buildSet {
            add(window.onset)
            add(window.endExclusive)
            clipped.forEach { add(it.onset); add(it.endExclusive) }
        }.sorted()
        val out = ArrayList<OwnerSegment>()
        var lastOwner = fallbackOwner
        for (i in 0 until seams.lastIndex) {
            val start = seams[i]
            val end = seams[i + 1]
            val owner = clipped.asSequence()
                .filter { start >= it.onset && start < it.endExclusive }
                .minWithOrNull(compareBy<OwnerCoverage> { it.priority }.thenBy { it.owner })
                ?.owner ?: lastOwner
            val previous = out.lastOrNull()
            if (previous != null && previous.owner == owner && previous.endExclusive == start) {
                out[out.lastIndex] = previous.copy(endExclusive = end)
            } else {
                out += OwnerSegment(owner, start, end)
            }
            lastOwner = owner
        }
        return out
    }

    /** Resolve the next observed onset, or [now] for the currently open physiological cycle. */
    fun cycleEnd(day: String, onsetByDay: Map<String, Long>, now: Long): Long {
        val onset = onsetByDay[day] ?: return now
        return onsetByDay.values.asSequence()
            .filter { it > onset && it <= now }
            .minOrNull()
            ?: now
    }

    /**
     * Sum locomotion deltas whose *later* sample lies in [onsetInclusive, endExclusive). Samples may
     * include one predecessor before onset so a counter increment crossing the boundary is attributed
     * correctly. Returns null only when there is not enough counter data to evaluate the cycle; an
     * observed, flat cycle is a real zero.
     */
    fun stepsInCycle(
        samples: List<StepSample>,
        onsetInclusive: Long,
        endExclusive: Long,
    ): Int? {
        if (endExclusive <= onsetInclusive) return 0
        val sorted = samples.sortedBy { it.ts }
        if (sorted.size < 2) return null
        val state = accumulateCyclePage(
            newCycleCount(StepsCounter.hasActivityClasses(sorted)),
            sorted,
            onsetInclusive,
            endExclusive,
        )
        return finishCycleCount(state)
    }

    fun newCycleCount(hasActivityClasses: Boolean): CycleCountState =
        CycleCountState(hasActivityClasses, previous = null, total = 0, evaluatedDelta = false)

    /** Fold one ascending page into [state]. Pages may overlap at their seam; duplicate timestamps skip. */
    fun accumulateCyclePage(
        state: CycleCountState,
        samples: List<StepSample>,
        onsetInclusive: Long,
        endExclusive: Long,
    ): CycleCountState {
        if (samples.isEmpty()) return state
        var previous = state.previous
        var total = state.total
        var evaluated = state.evaluatedDelta
        for (current in samples.sortedBy { it.ts }) {
            val prior = previous
            if (prior != null && current.ts <= prior.ts) continue
            if (prior != null && current.ts >= onsetInclusive && current.ts < endExclusive) {
                evaluated = true
                val delta = (current.counter - prior.counter) and 0xFFFF
                if (
                    StepsCounter.shouldCountDelta(current.activityClass, state.hasActivityClasses) &&
                    StepsCounter.isPlausibleDelta(prior.ts, current.ts, delta)
                ) {
                    total += delta.toLong()
                }
            }
            previous = current
        }
        return CycleCountState(state.hasActivityClasses, previous, total, evaluated)
    }

    fun finishCycleCount(state: CycleCountState): Int? =
        if (!state.evaluatedDelta) null else state.total.coerceAtMost(Int.MAX_VALUE.toLong()).toInt()
}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/PhysiologicalSteps.swift
import Foundation
import WhoopProtocol

public enum PhysiologicalSteps {
    public enum SleepKind: Sendable { case mainSleep, nap, unclassified }
    public struct SleepBlock: Sendable {
        public let onset: Int
        public let end: Int
        public let id: String
        public let editedOnset: Int?
        public let kind: SleepKind
        public var effectiveOnset: Int { editedOnset ?? onset }
        public init(onset: Int, end: Int, id: String? = nil, editedOnset: Int? = nil,
                    kind: SleepKind = .unclassified) {
            self.onset = onset; self.end = end; self.id = id ?? String(onset)
            self.editedOnset = editedOnset; self.kind = kind
        }
        fileprivate func withKind(_ kind: SleepKind) -> SleepBlock {
            SleepBlock(onset: onset, end: end, id: id, editedOnset: editedOnset, kind: kind)
        }
    }
    public struct CycleBoundary: Equatable, Sendable {
        public let sleepId: String
        public let onset: Int
        public init(sleepId: String, onset: Int) { self.sleepId = sleepId; self.onset = onset }
    }
    public struct CycleWindow: Equatable, Sendable {
        public let sleepId: String; public let onset: Int; public let endExclusive: Int
        public init(sleepId: String, onset: Int, endExclusive: Int) {
            self.sleepId = sleepId; self.onset = onset; self.endExclusive = endExclusive
        }
    }
    public struct OwnerSegment: Equatable, Sendable {
        public let owner: String; public let onset: Int; public let endExclusive: Int
        public init(owner: String, onset: Int, endExclusive: Int) {
            self.owner = owner; self.onset = onset; self.endExclusive = endExclusive
        }
    }
    public struct OwnerCoverage: Equatable, Sendable {
        public let owner: String; public let onset: Int; public let endExclusive: Int; public let priority: Int
        public init(owner: String, onset: Int, endExclusive: Int, priority: Int) {
            self.owner = owner; self.onset = onset; self.endExclusive = endExclusive; self.priority = priority
        }
    }
    private static let minMainSleepSeconds = 3 * 3_600

    public static func classifyForCycle(_ blocks: [SleepBlock], offsetSec: Int,
                                        habitualMidsleepSec: Int?) -> [SleepBlock] {
        guard !blocks.isEmpty else { return [] }
        let explicit = blocks.indices.filter { blocks[$0].kind == .mainSleep }
        let selectable = blocks.indices.filter { blocks[$0].kind != .nap }
        let selected: Set<Int>
        if !explicit.isEmpty {
            selected = Set(explicit)
        } else {
            let nightBlocks = selectable.map { SleepStageTotals.NightBlock(start: blocks[$0].effectiveOnset,
                                                                            end: blocks[$0].end) }
            let eligible = SleepStageTotals.bridgedNightGroups(nightBlocks, offsetSec: offsetSec)
                .filter { group in
                    let total = group.indices.reduce(0) { $0 + max(0, nightBlocks[$1].durationS) }
                    let onset = group.indices.map { nightBlocks[$0].start }.min()
                    return total >= minMainSleepSeconds && onset.map {
                        SleepStageTotals.isOvernightOnset($0, offsetSec: offsetSec)
                    } == true
                }
                .flatMap { $0.indices }
            let candidates = eligible.map { index in
                SleepStageTotals.NightBlock(start: blocks[selectable[index]].effectiveOnset,
                                            end: blocks[selectable[index]].end)
            }
            let picked = SleepStageTotals.mainNightGroupIndices(candidates, offsetSec: offsetSec,
                                                                habitualMidsleepSec: habitualMidsleepSec) ?? []
            selected = Set(picked.map { selectable[eligible[$0]] })
        }
        return blocks.indices.map { blocks[$0].withKind(selected.contains($0) ? .mainSleep : .nap) }
    }

    public static func cycleWindows(_ boundaries: [CycleBoundary], now: Int) -> [CycleWindow] {
        var ids = Set<String>()
        let ordered = boundaries.filter { $0.onset <= now && ids.insert($0.sleepId).inserted }
            .sorted { $0.onset < $1.onset }
        return ordered.indices.compactMap { index in
            let end = index + 1 < ordered.count ? ordered[index + 1].onset : now
            return end > ordered[index].onset
                ? CycleWindow(sleepId: ordered[index].sleepId, onset: ordered[index].onset, endExclusive: end)
                : nil
        }
    }

    /// Kotlin twin: `PhysiologicalSteps.ownerSegmentsFromCoverage`.
    public static func ownerSegmentsFromCoverage(_ window: CycleWindow, coverage: [OwnerCoverage],
                                                 fallbackOwner: String) -> [OwnerSegment] {
        guard window.endExclusive > window.onset else { return [] }
        let clipped = coverage.compactMap { item -> OwnerCoverage? in
            let start = max(window.onset, item.onset), end = min(window.endExclusive, item.endExclusive)
            return end > start ? OwnerCoverage(owner: item.owner, onset: start,
                endExclusive: end, priority: item.priority) : nil
        }
        let seams = Set([window.onset, window.endExclusive] + clipped.flatMap { [$0.onset, $0.endExclusive] }).sorted()
        var result: [OwnerSegment] = [], lastOwner = fallbackOwner
        for index in 0..<(seams.count - 1) {
            let start = seams[index], end = seams[index + 1]
            let owner = clipped.filter { start >= $0.onset && start < $0.endExclusive }
                .sorted { $0.priority == $1.priority ? $0.owner < $1.owner : $0.priority < $1.priority }
                .first?.owner ?? lastOwner
            if let last = result.last, last.owner == owner, last.endExclusive == start {
                result[result.count - 1] = OwnerSegment(owner: owner, onset: last.onset, endExclusive: end)
            } else {
                result.append(OwnerSegment(owner: owner, onset: start, endExclusive: end))
            }
            lastOwner = owner
        }
        return result
    }

    /// Same later-sample attribution, class gate and plausibility gate as the Kotlin twin.
    public static func stepsInCycle(_ samples: [StepSample], onsetInclusive: Int,
                                    endExclusive: Int, sleepSessions: [SleepSession]) -> Int? {
        guard endExclusive > onsetInclusive else { return 0 }
        let sorted = samples.sorted { $0.ts < $1.ts }
        let predecessor = sorted.last { $0.ts < onsetInclusive }
        let relevant = (predecessor.map { [$0] } ?? []) + sorted.filter {
            $0.ts >= onsetInclusive && $0.ts < endExclusive
        }
        guard relevant.count >= 2 else { return nil }
        let counted = SleepAwareStepCounter.count(relevant, sleepSessions: sleepSessions)
        var evaluated = false
        var total = 0
        let hasClasses = StepsCounter.hasActivityClasses(relevant)
        for index in 1..<relevant.count {
            let previous = relevant[index - 1], current = relevant[index]
            guard current.ts >= onsetInclusive, current.ts < endExclusive else { continue }
            evaluated = true
            let delta = (current.counter - previous.counter) & 0xffff
            guard StepsCounter.shouldCountDelta(activityClass: current.activityClass, hasActivityClasses: hasClasses),
                  StepsCounter.isPlausibleDelta(previousTs: previous.ts, currentTs: current.ts, delta: delta)
            else { continue }
            let inSleep = sleepSessions.contains { current.ts >= $0.start && current.ts < $0.end }
            if !inSleep { total += delta }
        }
        // SleepAware's accepted sleep/awake-gap contribution is window-scoped by the caller's samples.
        total += counted.acceptedAwakeGapTicks + counted.acceptedSleepBoutTicks
        return evaluated ? total : nil
    }
}
